File: car.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <unistd.h>

#include "car.h"

bool verbose = false;

static void usage(void)
{
	fprintf(stderr, "Usage: [-h] [-c] [-x] [file...]\n");
	exit(1);
}

int main(int argc, char *argv[])
{
	int c;
	bool comp = false, extr = false;
	char *file = NULL;

	while ((c = getopt(argc, argv, "hcx:f:v")) != -1) {
		switch(c) {
		case 'h':
			usage();
			break;
		case 'c':
			comp = true;
			break;
		case 'x':
			extr = true;
			break;
		case 'f':
			file = optarg;
			break;
		case 'v':
			verbose = true;
			break;
		case '?':
		default:
			usage();
		}
	}

	if (!comp && !extr) {
		fprintf(stderr, "Error: must specify either -c or -x\n");
		return 1;
	}

	if (comp && extr) {
		fprintf(stderr, "Error: cannot specify both -c and -x\n");
		return 1;
	}

	if (optind == argc) {
		fprintf(stderr, "Error: must specify at least one file\n");
		return 1;
	}

	if (comp) {
		printf("Compressing %s to %s\n", argv[optind], file);
		compress(argv[optind], file);
	} else {
		printf("Extracting %s to %s\n", file, argv[optind]);
		extract(file, argv[optind]);
	}
}

File: car.go
package main

import (
	"flag"
	"fmt"
	"os"
)

func b(b bool) int {
	if b {
		return 1
	}
	return 0
}

func main() {
	var err error

	t := flag.Bool("t", false, "list")
	c := flag.Bool("c", false, "archive")
	x := flag.Bool("x", false, "extract")
	file := flag.String("f", "", "file")
	verbose := flag.Bool("v", false, "verbose")
	flag.Parse()

	if b(*t)+b(*c)+b(*x) != 1 {
		fmt.Fprintln(os.Stderr, "Exactly one ption -t, -c or -x must be specified")
		os.Exit(1)
	}

	var a archive = &car{
		verbose: *verbose,
		list:    *t,
	}

	switch {
	case *c:
		if flag.NArg() == 0 {
			fmt.Fprintln(os.Stderr, "Missing path to compress")
			os.Exit(1)
		}

		err = a.archive(flag.Args(), *file)

	case *t, *x:
		err = a.extract(*file)
	}

	if err != nil {
		fmt.Fprintln(os.Stderr, err)
	}
	os.Exit(a.(*car).error)
}
